lib/charms/identity_platform_login_ui_operator/v0/login_ui_endpoints.py

```diff
@@ -0,0 +1,186 @@
+            return
+
+        relations = self.model.relations[self._relation_name]
+        for relation in relations:
+            relation.data[self._charm.app].update(
+                {
+                    "consent_url": f"{endpoint}/consent",
+                    "error_url": f"{endpoint}/error",
+                    "index_url": f"{endpoint}/index",
+                    "login_url": f"{endpoint}/login",
+                    "oidc_error_url": f"{endpoint}/oidc_error",
+                    "registration_url": f"{endpoint}/registration",
+                    "default_url": endpoint,
+                }
+            )
+
+
+class LoginUIEndpointsRelationError(Exception):
+    """Base class for the relation exceptions."""
+
+    pass
+
+
+class LoginUIEndpointsRelationMissingError(LoginUIEndpointsRelationError):
+    """Raised when the relation is missing."""
+
+    def __init__(self) -> None:
+        self.message = "Missing ui-endpoint-info relation with Identity Platform Login UI"
+        super().__init__(self.message)
+
+
+class LoginUIEndpointsRelationDataMissingError(LoginUIEndpointsRelationError):
+    """Raised when information is missing from the relation."""
+
+    def __init__(self, message: str) -> None:
+        self.message = message
+        super().__init__(self.message)
+
+
+class LoginUIEndpointsRelationUnavailableError(LoginUIEndpointsRelationError):
+    """Raised when Login UI cannot be accessed."""
+
+    def __init__(self, message: str) -> None:
+        self.message = message
+        super().__init__(self.message)
+
+
+class LoginUIEndpointsRequirer(Object):
+    """Requirer side of the ui-endpoint-info relation."""
+
+    def __init__(self, charm: CharmBase, relation_name: str = RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self._relation_name = relation_name
+
+    def get_login_ui_endpoints(self) -> Optional[Dict]:
+        """Get the Identity Platform Login UI endpoints."""
+        if not self.model.unit.is_leader():
+            return None
+        endpoints = self.model.relations[self._relation_name]
+        if len(endpoints) == 0:
+            raise LoginUIEndpointsRelationMissingError()
+
+        data = endpoints[0].data[endpoints[0].app]
+        return_dict = {}
+
+        for k in RELATION_KEYS:
+            if k not in data:
+                raise LoginUIEndpointsRelationDataMissingError(
+                    "Missing endpoints in ui-endpoint-info relation data"
+                )
+            if data[k] == "":
+                raise LoginUIEndpointsRelationDataMissingError(
+                    "Missing endpoints in ui-endpoint-info relation data"
+                )
+            return_dict[k] = data[k]
+        
+        if return_dict["default_url"] == "":
+            raise LoginUIEndpointsRelationUnavailableError(
+                "Endpoints in ui-endpoint-info unavailable"
+            )
+
+        return return_dict
```

Comment:
I think you can simplify this, with:
```suggestion
        if any(not data.get(k := key) for key in RELATION_KEYS):
            raise LoginUIEndpointsRelationDataMissingError(
                f"Missing endpoint {k} in ui-endpoint-info relation data"
            )
        return dict(data)
```

Reply:
done